dbcon/mysql/ha_mcs_pushdown.cpp

```diff
@@ -792,71 +781,184 @@ create_columnstore_select_handler(THD* thd, SELECT_LEX* select_lex)
     // or unsupported feature.
-    handler = new ha_columnstore_select_handler(thd, select_lex);
+    ha_columnstore_select_handler* handler =
+        new ha_columnstore_select_handler(thd, select_lex);
+
     JOIN *join = select_lex->join;
-    bool unsupported_feature = false;
+
+    if (select_lex->first_cond_optimization &&
+        select_lex->handle_derived(thd->lex, DT_MERGE))
+    {
+        if (!thd->is_error())
+        {
+            my_printf_error(ER_INTERNAL_ERROR, "%s", MYF(0),
+                "Error occured in select_lex::handle_derived()");
+        }
+
+        return handler;
+    }
+
+    // This is partially taken from JOIN::optimize_inner() in sql/sql_select.cc
+    if (select_lex->first_cond_optimization)
     {
+        create_explain_query_if_not_exists(thd->lex, thd->mem_root);
         Query_arena *arena, backup;
         arena = thd->activate_stmt_arena_if_needed(&backup);
-        disable_indices_for_CEJ(thd);
+        COND* conds = join->conds;
+        select_lex->where = conds;
+
+        if (isPS)
+        {
+            select_lex->prep_where = conds ? conds->copy_andor_structure(thd) : 0;
+        }
+
+        select_lex->update_used_tables();
 
         if (arena)
             thd->restore_active_arena(arena, &backup);
 
-        if (select_lex->handle_derived(thd->lex, DT_MERGE))
+#ifdef DEBUG_WALK_COND
+        if (conds)
         {
-            unsupported_feature = true;
-            handler->err_msg.assign("create_columnstore_select_handler(): \
-                Internal error occured in SL::handle_derived()");
+            conds->traverse_cond(cal_impl_if::debug_walk, NULL, Item::POSTFIX);
         }
+#endif
+    }
 
-        COND *conds = nullptr;
-        if (!unsupported_feature)
+    // Attempt to execute the query using the select handler.
+    // If query execution fails and columnstore_select_handler=AUTO,
+    // fallback to table mode.
+    // Skip execution for EXPLAIN queries
+    if (!thd->lex->describe)
+    {
+        // This is taken from JOIN::optimize()
+        join->fields= &select_lex->item_list;
+
+        // Instantiate handler::table, which is the place for the result set.
+        if (handler->prepare())
         {
-            // Rewrite once for PS
-	    // Refer to JOIN::optimize_inner() in sql/sql_select.cc
-	    // for details on the optimizations performed in this block.
-            if (select_lex->first_cond_optimization)
+            // check fallback
+            if (select_handler_mode == 2) // == 2 means columnstore_select_handler=AUTO
             {
-                create_explain_query_if_not_exists(thd->lex, thd->mem_root);
-                arena = thd->activate_stmt_arena_if_needed(&backup);
-                select_lex->first_cond_optimization= false;
-                conds = join->conds;
-                select_lex->where = conds;
+                push_warning(thd, Sql_condition::WARN_LEVEL_WARN, 9999,
```

Comment:
The warning here might produce an error to the client depending on the sql_mode session variable value. So the behavior here IMHO an open question whether to log it into our log or return a warning over a client proto.

Reply:
It does not produce an error irrespective of the sql_mode since we have already called thd->clear_error().

Reply:
My point is that certain sql_mode treats warning as an error so that this behavior might be not very MODE_STRICT_TABLE(sql_mode value) freiendly.